server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/CreateConditionAction.java

```diff
@@ -22,46 +22,104 @@
+    this.userSession = userSession;
+    this.dbClient = dbClient;
+    this.qualityGateConditionsUpdater = qualityGateConditionsUpdater;
   }
 
   @Override
   public void define(WebService.NewController controller) {
-    WebService.NewAction createCondition = controller.createAction("create_condition")
+    WebService.NewAction createCondition = controller.createAction(ACTION_CREATE_CONDITION)
       .setDescription("Add a new condition to a quality gate. Require Administer Quality Gates permission")
       .setPost(true)
       .setSince("4.3")
       .setHandler(this);
 
     createCondition
-      .createParam(QualityGatesWsParameters.PARAM_GATE_ID)
+      .createParam(PARAM_GATE_ID)
       .setDescription("ID of the quality gate")
       .setRequired(true)
       .setExampleValue("1");
 
-    QualityGatesWs.addConditionParams(createCondition);
+    addConditionParams(createCondition);
   }
 
   @Override
   public void handle(Request request, Response response) {
-    QualityGatesWs.writeQualityGateCondition(
-      qualityGates.createCondition(
-        QualityGatesWs.parseId(request, QualityGatesWsParameters.PARAM_GATE_ID),
-        request.mandatoryParam(QualityGatesWsParameters.PARAM_METRIC),
-        request.mandatoryParam(QualityGatesWsParameters.PARAM_OPERATOR),
-        request.param(QualityGatesWsParameters.PARAM_WARNING),
-        request.param(QualityGatesWsParameters.PARAM_ERROR),
-        request.paramAsInt(QualityGatesWsParameters.PARAM_PERIOD)
-        ), response.newJsonWriter()
-      ).close();
+    userSession.checkPermission(QUALITY_GATE_ADMIN);
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      writeProtobuf(doHandle(toWsRequest(request), dbSession), request, response);
+      dbSession.commit();
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private CreateConditionWsResponse doHandle(CreateConditionRequest request, DbSession dbSession){
+    QualityGateConditionDto condition = qualityGateConditionsUpdater.createCondition(dbSession,
+      request.getQualityGateId(),
+      request.getMetricKey(),
+      request.getOperator(),
+      request.getWarning(),
+      request.getError(),
+      request.getPeriod());
+
+    CreateConditionWsResponse.Builder response = CreateConditionWsResponse.newBuilder()
+      .setId(condition.getId())
+      .setMetric(condition.getMetricKey())
+      .setOp(condition.getOperator());
+    String warning = condition.getWarningThreshold();
+    if (warning != null) {
+      response.setWarning(warning);
+    }
+    String error = condition.getErrorThreshold();
+    if (error != null) {
+      response.setError(error);
+    }
+    Integer period = condition.getPeriod();
+    if (period != null) {
+      response.setPeriod(period);
+    }
+    return response.build();
+  }
+
+  private static CreateConditionRequest toWsRequest(Request request) {
```

Comment:
if what we want is to embrace more and more the functional style of programming (I didn't say functional programming)